File: client/Java/Juno/juno-client-api/src/main/java/com/paypal/juno/client/io/OperationStatus.java
//  
//  Copyright 2023 PayPal Inc.
//  
//  Licensed to the Apache Software Foundation (ASF) under one or more
//  contributor license agreements.  See the NOTICE file distributed with
//  this work for additional information regarding copyright ownership.
//  The ASF licenses this file to You under the Apache License, Version 2.0
//  (the "License"); you may not use this file except in compliance with
//  the License.  You may obtain a copy of the License at
//  
//     http://www.apache.org/licenses/LICENSE-2.0
//  
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//  
package com.paypal.juno.client.io;

import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;

public enum OperationStatus {

    // Client side Response codes, Error Text status success(true) and failure(false)

	Success(0,"No error",true),
	NoKey(1,"Key not found",true),
	BadParam(2,"Bad parameter",false),
	UniqueKeyViolation(3,"Duplicate key",true),
	RecordLocked(4,"Record Locked",true),
	IllegalArgument(5,"Illegal argument",false),
	ConditionViolation(6,"Condition in the request violated",true),
	InternalError(7,"Internal error",false),
	QueueFull(8,"Outbound client queue full",false),
	NoStorage(9,"No storage server running",false),
	TTLExtendFailure(10,"Failure to extend TTL on get",true),
	ResponseTimeout(11,"Response Timed out",false),
	ConnectionError(12,"Connection Error",false),
	UnknownError(13,"Unknown Error",false);

	
	private final int code;
	private final String errorText;
	private final boolean txnOk;

	private static final Map<Integer, OperationStatus> lookup = new HashMap<Integer, OperationStatus>();

	static {
		for (OperationStatus s : EnumSet
				.allOf(OperationStatus.class))
			lookup.put(s.getCode(), s);
	}

	/**
	 * Constructor
	 * 
	 * @param code
	 * @param errorText
	 */
	OperationStatus(int code, String errorText,boolean txnOk) {
		this.code = code;
		this.errorText = errorText;
		this.txnOk = txnOk;
	}

	public int getCode() {
		return this.code;
	}

	public String getErrorText() {
		return this.errorText;
	}

	public static OperationStatus get(int code) {
		return lookup.get(code);
	}
	
	public boolean isTxnOk(){
		return txnOk;
	}
	
}

File: client/Ruby/juno/lib/juno/Client/operation_status.rb
# frozen_string_literal: true

# Constant for Operation Status in JunoResponse
module Juno
  module Client
    class OperationStatus
      SUCCESS = { code: 0, error_msg: 'No error', txnOk: true }.freeze
      NO_KEY = { code: 1, error_msg: 'Key not found', txnOk: true }.freeze
      BAD_PARAM = { code: 2, error_msg: 'Bad parameter', txnOk: false }.freeze
      UNIQUE_KEY_VIOLATION = { code: 3, error_msg: 'Duplicate key', txnOk: true }.freeze
      RECORD_LOCKED = { code: 4, error_msg: 'Record Locked', txnOk: true }.freeze
      ILLEGAL_ARGUMENT = { code: 5, error_msg: 'Illegal argument', txnOk: false }.freeze
      CONDITION_VIOLATION = { code: 6, error_msg: 'Condition in the request violated', txnOk: true }.freeze
      INTERNAL_ERROR = { code: 7, error_msg: 'Internal error', txnOk: false }.freeze
      QUEUE_FULL = { code: 8, error_msg: 'Outbound client queue full', txnOk: false }.freeze
      NO_STORAGE = { code: 9, error_msg: 'No storage server running', txnOk: false }.freeze
      TTL_EXTEND_FAILURE = { code: 10, error_msg: 'Failure to extend TTL on get', txnOk: true }.freeze
      RESPONSE_TIMEOUT = { code: 11, error_msg: 'Response Timed out', txnOk: false }.freeze
      CONNECTION_ERROR = { code: 12, error_msg: 'Connection Error', txnOk: false }.freeze
      UNKNOWN_ERROR = { code: 13, error_msg: 'Unknown Error', txnOk: false }.freeze

      @@status_code_map = nil

      def self.initialize_map
        @@status_code_map = {}

        constants.each do |const|
          const_obj = const_get(const)
          @@status_code_map[const_obj[:code]] = const_obj
        end
      end

      def self.get(status_code)
        initialize_map if @@status_code_map.nil?
        return @@status_code_map[status_code] if @@status_code_map.key?(status_code)

        INTERNAL_ERROR
      end
    end
  end
end
